### PR TITLE
add beforeRestore/afterRestore hooks

### PR DESCRIPTION
**Pull request by rodrigoapereira**
For this issue [#4371](https://github.com/sequelize/sequelize/issues/4371)

### CODE REVIEW

File: changelog.md [hunk 1]
```diff
@@ -1,3 +1,6 @@
+# Next
+  [ADDED] before/after and beforeBulk/afterBulk restore hooks [#4371](https://github.com/sequelize/sequelize/issues/4371)
```
Comment (janmeier) [inline changelog.md]: beforeRestore/afterRestore ;)

File: test/integration/hooks.test.js [hunk 2]
```diff
@@ -19,7 +19,19 @@ describe(Support.getTestDialectTeaser('Hooks'), function() {
...
+      username: DataTypes.STRING,
+      mood: {
+        type: DataTypes.ENUM,
+        values: ['happy', 'sad', 'neutral']
+      }
+    }, {
+      paranoid: true
+    });
+
+    return this.User.sync({ force: true }).bind(this).then(function() {
```
Comment (janmeier) [inline test/integration/hooks.test.js]: `return this.sequelize.sync(..`

File: test/integration/hooks.test.js [hunk 3]
```diff
@@ -667,6 +742,146 @@ describe(Support.getTestDialectTeaser('Hooks'), function() {
...
+        }).then(function() {
+          expect(beforeBulk).to.be.true;
+          expect(afterBulk).to.be.true;
+          expect(beforeHook).to.be.true;
+          expect(afterHook).to.be.true;
+        });
+      });
+
+      it('should run the after/before functions for each item restored with an error', function() {
+        var self = this
```
Comment (janmeier) [inline test/integration/hooks.test.js]: Would you mind using spies instead? 

```js
beforeBulk = sinon.spy();
this.ParanoidUser.beforeBulkRestore(beforeBulk)
```

**Comment by janmeier** [comment]
Looks good, just a few minor comments

**Comment by rodrigoapereira** [comment]
Thanks for reply @janmeier.
I added a new commit for fix pull request comments.

**Comment by janmeier** [comment]
Looking good, please rebase and squash into a single, descriptive commit then we're good to go

**Comment by rodrigoapereira** [comment]
Ok, i'm finished the squash.

**Comment by janmeier** [comment]
Thanks! :)

**Pull request merged**